ctru-rs/src/services/soc.rs

```diff
@@ -31,20 +31,28 @@ impl Soc {
     ///
     /// This function will return an error if the `Soc` service is already initialized
     pub fn init_with_buffer_size(num_bytes: usize) -> crate::Result<Self> {
-        match SOC_ACTIVE.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst) {
-            Ok(_) => unsafe {
-                let soc_mem = memalign(0x1000, num_bytes) as *mut u32;
-
-                let r = socInit(soc_mem, num_bytes as u32);
+        let _service_handler = ServiceHandler::new(
+            &SOC_ACTIVE,
+            true,
+            || {
+                let soc_mem = unsafe { memalign(0x1000, num_bytes) } as *mut u32;
+                let r = unsafe { ctru_sys::socInit(soc_mem, num_bytes as u32) };
                 if r < 0 {
-                    free(soc_mem as *mut _);
-                    Err(r.into())
-                } else {
-                    Ok(Self { soc_mem })
+                    return Err(r.into());
                 }
+
+                Ok(())
+            },
+            // `socExit` returns an error code. There is no documentantion of when errors could happen,
+            // but we wouldn't be able to handle them in the `Drop` implementation anyways.
+            // Surely nothing bad will happens :D
```

Comment:
We could panic in the drop, but it seems a little over the top.

Reply:
I’d avoid panicking. Also, if the code works properly (and thus the service is actually open when calling the `Drop`) then I believe there aren’t any problems.